packages/hooks/src/useRoute.ts

```diff
@@ -1,158 +1,58 @@
+		(api: CoreAPI, route: string, params: Record<string, ModelValue>) => {
+			setIsLoading(true)
+			api
+				.getRoute<T>(route, params)
+				.then((results) => setData(results))
+				.catch((err) => setError(err))
+				.finally(() => setIsLoading(false))
 		},
-		[callback, JSON.stringify(params)]
+		500,
+		{ leading: true, trailing: true }
 	)
 
+	const routeRef = useRef<string | null>(null)
+	const paramsRef = useRef<Record<string, ModelValue>>({})
+
 	useEffect(() => {
-		if (applicationData === null || ws === null) {
-			return
-		} else if (!applicationData.routes.includes(route)) {
-			setError(new Error(`${applicationData.uri} has no route ${JSON.stringify(route)}`))
-			setData(null)
-			setIsLoading(false)
-			if (callback) setTimeout(() => callback(null, error))
+		if (api === null) {
 			return
 		}
 
-		setIsLoading(true)
-
-		if (subscribe) {
-			if (ws.readyState === ws.OPEN) {
-				// console.log("ws: subscribing", url)
-				ws.addEventListener("message", listener)
-				ws.send(JSON.stringify({ action: "subscribe", data: { route, params } }))
-
-				return () => {
-					// console.log("ws: unsubscribing", url)
-					ws.removeEventListener("message", listener)
-					if (ws.readyState === ws.OPEN) {
-						ws.send(JSON.stringify({ action: "unsubscribe", data: { route, params } }))
-					}
-				}
-			}
-
-			// const source = new EventSource(url)
-			// source.onmessage = (message: MessageEvent<string>) => {
-			// 	const data = JSON.parse(message.data)
-			// 	setData(data)
-			// 	setIsLoading(false)
-			//  if (callback) callback(message.data, null)
-			// }
-
-			// source.onerror = (event) => {
-			// 	console.warn("Connection error in EventSource subscription")
-			// 	console.warn(event)
-			// 	setIsLoading(true)
-			//  if (callback) callback(null, new Error("Connection error in EventSource subscription"))
-			// }
+		if (route !== routeRef.current || !compareObjects(params, paramsRef.current)) {
+			refetch(api, route, params)
+		}
 
-			// const handleBeforeUnload = () => source.close()
-			// window.addEventListener("beforeunload", handleBeforeUnload)
+		routeRef.current = route
+		paramsRef.current = params
 
-			// return () => {
-			// 	window.removeEventListener("beforeunload", handleBeforeUnload)
-			// 	source.close()
-			// }
-		} else {
-			fetch(url)
-				.then((res) => res.json())
-				.then((data) => {
-					setError(null)
-					setData(data)
-					setIsLoading(false)
-					if (callback) setTimeout(() => callback(data, null))
-				})
-				.catch((err) => {
-					setError(err)
-					setData(null)
-					setIsLoading(false)
-					if (callback) setTimeout(() => callback(null, err))
-				})
+		if (subscribe) {
+			const listener = ({}: Event) => refetch(api, route, params)
+			api.addEventListener("update", listener)
+			return () => api.removeEventListener("update", listener)
 		}
-	}, [route, url, !!applicationData, subscribe, ws, readyState, callback])
+	}, [api, route, params])
```

Comment:
Is it possible a value in params changes, but `params` itself doesnt?

Reply:
oh hmm hadn't thought of that... it's possible with React hooks I feel like the expectation is that it refreshes only if the object itself changes